internal/cloudapi/v2/search.go

```diff
@@ -0,0 +1,85 @@
+package v2
+
+// SearchRequest methods
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/osbuild/images/pkg/distrofactory"
+	"github.com/osbuild/images/pkg/reporegistry"
+	"github.com/osbuild/images/pkg/rpmmd"
+	"github.com/osbuild/osbuild-composer/internal/worker"
+)
+
+func (request *SearchRequest) Search(df *distrofactory.Factory, rr *reporegistry.RepoRegistry, workers *worker.Server) (rpmmd.PackageList, error) {
+	distro := df.GetDistro(request.Distribution)
+	if distro == nil {
+		return nil, HTTPError(ErrorUnsupportedDistribution)
+	}
+	distroArch, err := distro.GetArch(request.Architecture)
+	if err != nil {
+		return nil, HTTPErrorWithInternal(ErrorUnsupportedArchitecture, err)
+	}
+
+	var repos []rpmmd.RepoConfig
+	if request.Repositories != nil {
+		repos, err = convertRepos(*request.Repositories, []Repository{}, []string{})
+		if err != nil {
+			// Error comes from genRepoConfig and is already an HTTPError
+			return nil, err
+		}
+	} else {
+		repos, err = rr.ReposByArchName(request.Distribution, distroArch.Name(), false)
+		if err != nil {
+			return nil, HTTPErrorWithInternal(ErrorInvalidRepository, err)
+		}
+	}
+
+	// Send the search request to the worker
+	searchJobID, err := workers.EnqueueSearch(&worker.SearchJob{
+		Packages:         request.Packages,
+		Repositories:     repos,
+		ModulePlatformID: distro.ModulePlatformID(),
+		Arch:             distroArch.Name(),
+		Releasever:       distro.Releasever(),
+	}, "")
+	if err != nil {
+		return nil, HTTPErrorWithInternal(ErrorEnqueueingJob, err)
+	}
+
+	// Limit how long a search can take
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*depsolveTimeoutMin)
+	defer cancel()
+
+	// Wait until search job is finished, fails, or is canceled
+	var result worker.SearchJobResult
+	for {
+		time.Sleep(time.Millisecond * 50)
+		info, err := workers.SearchJobInfo(searchJobID, &result)
+		if err != nil {
+			return nil, HTTPErrorWithInternal(ErrorFailedToDepsolve, err)
+		}
+		if result.JobError != nil {
+			return nil, HTTPErrorWithInternal(ErrorFailedToDepsolve, err)
+		}
+		if info.JobStatus != nil {
+			if info.JobStatus.Canceled {
+				return nil, HTTPErrorWithInternal(ErrorFailedToDepsolve, err)
+			}
+
+			if !info.JobStatus.Finished.IsZero() {
+				break
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil, HTTPErrorWithInternal(ErrorFailedToDepsolve, fmt.Errorf("Search job %q timed out", searchJobID))
+		default:
+		}
+	}
```

Comment:
I'll let other REST API experts comment, but I'm unsure if this POST should wait on the result instead of returning an operation UUID and expecting the caller to GET the result. But maybe this is acceptable. We may end up waiting on many package searches in goroutines, which will take non-trivial amount of time.

Reply:
Making the front end have to poll makes things harder. In general once the system is 'warmed up' the depsolve and search should return fairly quickly.

Reply:
@croissanne can correct me if I'm wrong. I think the POST/GET would be the technically correct way of doing things. But I think there are examples of this pattern in the service already (POST with the result).

Polling with a GET request in the frontend does sounds like a bit of a pain.